### ISSUE TITLE
Investigate non-deterministic numerical instability in 9 bodies scenario

### ISSUE THREAD

**Issue by ezocher** [ezocher/Gravity-Sandbox#2]
In the 9 bodies scenario the central body should never move. Because of cumulative numerical errors (confirm) it eventually does.

Even when there are errors, the simulation runs should give identical results since the exact same calculations are run in the exact same order with no randomness introduced. I ran three runs of the 9 bodies scenario and got three totally different paths for the central body which diverged from 0, 0 at different iterations. _WTF?_ 

Screenshots are here: https://github.com/ezocher/Gravity-Sandbox/blob/master/Images/
NineBodies-NumericalInstability1.PNG (and 2 and 3)

**Comment by ezocher** [comment]
The instability is now happening much earlier in the scenario run since I implemented the calculation precision enhancement in issue #3.

The instability is pretty repeatable, occurring about 40 seconds into a run with calculations per frame at 200.

**Comment by ezocher** [comment]
With the calculations per frame at 500 it happens repeatably at about 72 seconds. The really odd thing is that all nine bodies end up travelling pretty rapidly to the left. I think that the potential energy in the system at the start is symmetrical about the center point of the simulation, so this doesn't make sense to me from a physics perspective.

**Comment by ezocher** [comment]
With the calculations per frame at 100 it happens repeatably at about 40 seconds and everything heads upward.

Is this a cumulative error, or a specific event? Center body starts having anomalous velocity in y at about 27 seconds and then this starts to rapidly increase over the next 10 seconds.

**Try logging calculation details to a text file and then analyze in Excel.** Use Show trails to turn logging on and off. Accumulate values in memory from a Run click to a Pause click and then dump them to a text file.

**Comment by ezocher** [comment]
- [x] The ISS seems to be slowing down and moving higher every orbit, characterize and investigate cause

**Comment by ezocher** [comment]
Checked off the ISS issue above and moved it to a new issue (issue #37), as it's probably a different root cause (truncation?)

**Comment by ezocher** [comment]
The symmetrical scenarios that break out of symmetry were caused by (a + b + a) not being equal to (a + a + b) for certain values of a and b with similar magnitudes. These one bit roundoff errors get magnified over many calculation cycles (and new errors are introduced) leading to eventual macro breakdown of symmetry.  See FloatingPointNotAssociative.cs in NumericTypesTests.sln for the specific trigger I found in the 5 bodies scenario. Also see the descriptions of Roundoff Error and Stability on pages 10-11 of _Numerical Recipes, 3e_.

This is fixed by rounding off (to achieve masking) the repeating calculations so that the errors in the lowest significant bits of the mantissa don't appear in the calculations. Fixed in numerical-investigations branch.
